06x-new-kafka-roller.md

```diff
@@ -0,0 +1,220 @@
+   - If a node has configuration changes but they cannot be dynamically updated, add nodes into the `RESTART` group.
+   - If a node has no configuration changes, put the node into the `NOP` group.
+
+7. Reconfigure each node in `RECONFIGURE` group:
+   - If `numReconfigs` of a node is greater than the configured `maxReconfigs`, add a restart reason to its context. Otherwise continue.
+   - Send `incrementalAlterConfig` request with its config updates.
+   - Transitions the node's state to `RECONFIGURED` and increment its `numReconfigs`.
+   - Wait for each node that got configurations updated until they have `SERVING` within the `postOperationalTimeoutMs`. 
+   - If the `postOperationalTimeoutMs` is reached, restart from step 3.
+
+8. If at this point, the `RESTART` group is empty, the reconciliation will be completed successfully.
+
+9. Otherwise, batch nodes in `RESTART` group and get the next batch to restart:
+   - Further categorize nodes based on their roles so that the following restart order can be enforced:
+       1. `NON_ACTIVE_PURE_CONTROLLER` - Pure controller that is not the active controller
+       2. `ACTIVE_PURE_CONTROLLER` - Pure controller that is the active controller (the quorum leader)
+       3. `BROKER_AND_NOT_ACTIVE_CONTROLLER` - Node that is at least a broker but also might be a controller (combined) and is not the active controller
+       4. `BROKER_AND_ACTIVE_CONTROLLER` - Combined node that is the active controller (the quorum leader)
+
+       The returned batch will contain only one node if it is not `BROKER_AND_NOT_ACTIVE_CONTROLLER` group, so that controllers are restarted one at a time.
+  
+   - If `NON_ACTIVE_PURE_CONTROLLER` group is non empty, return the first node that can be restarted without impacting the quorum health (more on this later).
+   - If `ACTIVE_PURE_CONTROLLER` group is non empty, return the node if it can be restarted without impacting the quorum health. Otherwise return an empty set.
+   - If `BROKER_AND_NOT_ACTIVE_CONTROLLER` group is non empty, batch the broker nodes:
+       - remove the node from the list, if it is a combined node and cannot be restarted without impacting the quorum health so that it does get included in a batch
+       - build a map of nodes and their replicating partitions by sending describeTopics request to Admin API
+       - batch the nodes that do not have any partitions in common therefore can be restarted together
+       - remove nodes that have an impact on the availability from the batches (more on this later)
+       - return the largest batch
+   - If an empty batch is returned, that means none of the nodes met the safety conditions such as availability and qourum health impact. In this case, check their `numAttempts` and if any of them is equal to or greater than `maxAttempts`, throw `UnrestartableNodesException`. Otherwise increment their `numAttempts` and restart from step 3.
```

Comment:
I think when you're describing these things you should provide the _rationale_ for _why_ we've chosen the algorithm to guarantee certain properties, like that controllers are only ever restarted one at a time, but that we consider it's fine to restart brokers in parallel. You could use a blockquote so that such comments about the algorithm were visually distinct from the algorithm itself.